File: we-persistence-file/src/main/java/de/percsi/webEngine/persistence/file/tools/TextFileCRUD.java
package de.percsi.webEngine.persistence.file.tools;


import io.vavr.control.Option;
import io.vavr.control.Try;
import lombok.Value;
import lombok.extern.slf4j.Slf4j;

import java.nio.charset.StandardCharsets;
import java.nio.file.*;

@Slf4j

public final class TextFileCRUD {

  private final Path path;

  TextFileCRUD(Path path) {
    this.path = path;
  }


  Option<String> create(String data) {
    return Try.of(() ->{
      Files.createFile(path);
      Files.write(path,data.getBytes(StandardCharsets.UTF_8));
      return Option.of(data);
    }).getOrElse(Option.none());
  }

  Option<String> read() {
    return Try.withResources(()-> Files.lines(path, StandardCharsets.UTF_8))
        .of(stringStream -> {
          StringBuilder sb = new StringBuilder();
          stringStream.forEach(strLine -> sb.append(strLine).append("\n"));
          return Option.of(sb.toString());
        }).getOrElse(Option.none());
  }

  Option<String> update(String data) {
    return Try.of(() ->{
      if (Files.isWritable(path)) {
        Files.write(path,data.getBytes(StandardCharsets.UTF_8), StandardOpenOption.WRITE);
        return Option.of(data);
      } else {
        throw new NoSuchFileException(path.toString());
      }
    }).getOrElse(Option.none());
  }

  boolean delete(String fileName) {
    return Try.of(() -> {
      Files.delete(path);
      return true;
    }).getOrElse(false);
  }

}

File: we-persistence-file/src/main/kotlin/de/percsi/webEngine/persistence/file/tools/TextFileCRUD.kt
package de.percsi.webEngine.persistence.file.tools
import org.slf4j.LoggerFactory
import java.nio.charset.StandardCharsets
import java.nio.file.*
import kotlin.streams.toList

class TextFileCRUD constructor(private val path: Path) {

    private val log = LoggerFactory.getLogger(TextFileCRUD::class.java)

    fun create(data: String): String? {
        return try {
            Files.write(path,data.toByteArray(StandardCharsets.UTF_8),StandardOpenOption.CREATE_NEW)
            log.info("File is created (${path})")
            data
        } catch (e: Exception) {
            log.error("File could not be created (${path}): ${e.localizedMessage}")
            null
        }
    }

    fun read(): String? {
        return try {
            val data = Files.lines(path).map { "$it\n" }.toList().joinToString(separator = "")
            log.info("File is read (${path})")
            data
        } catch (e: Exception) {
            log.error("File could not be read ($path): ${e.localizedMessage}")
            null
        }
    }

    fun update(data: String): String? {
        return try {
            return if (Files.isWritable(path)) {
                Files.write(path, data.toByteArray(StandardCharsets.UTF_8), StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING)
                log.info("File is updated ($path)")
                data
            }
            else
                throw NoSuchFileException("File does not exists ")
        } catch (e: Exception) {
            log.error("File could not be updated($path): ${e.localizedMessage}")
            null
        }
    }

    fun delete(): Boolean {
        return try {
            Files.delete(path)
            true
        } catch (e: Exception) {
            log.warn("File could not be deleted ($path): ${e.localizedMessage}")
            false
        }
    }
}